### PR TITLE
feat: Add multiple sme management endpoints

### PR DESCRIPTION
**Pull request by oyekanmiayo**
- [x] SME Users can add SME Projects
- [x] SME Users can see only SME Projects tied to their SME
- [x] SME Users can add milestones to SME Projects
- [x] SME Users can view milestones for an SME Project
- [x] SME User can upload documents for Milestones - Add permission to ensure that SME Users can only upload
- [x] Investor Users (only) can see a list of all SME Projects
- [x] SME User cannot fetch SME Projects that don’t have Milestones
- [x] SME User should be able to update the status of a Milestone

## How Has This Been Tested?
Multiple unit tests have been written.

### CODE REVIEW

File: sme_management/tests/test_sme_apis.py [hunk 1]
```diff
@@ -0,0 +1,345 @@
+import os
+
+from django.test import TestCase
+from django.urls import reverse
+from django.contrib.auth import get_user_model
+from django.conf import settings
+from django.core.files import File
```
Comment (oyekanmiayo) [inline sme_management/tests/test_sme_apis.py]: Codacy Issue found: [Unused File imported from django.core.files](https://app.codacy.com/gh/BuildForSDG/Team-004-Backend/pullRequest?prid=5635476)

**Comment by oyekanmiayo** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 8
           

Complexity increasing per file
==============================
- sme_management/migrations/0014_auto_20200531_0841.py  1
- sme_management/migrations/0018_auto_20200601_1018.py  1
- sme_management/migrations/0017_auto_20200601_1010.py  1
- sme_management/migrations/0015_smeprojectmilestones_proof_of_completion.py  1
- investor_management/tests/test_investor_apis.py  1
- sme_management/migrations/0016_smeprojectmilestones_status.py  1
- sme_management/migrations/0013_auto_20200531_0747.py  1
- sme_management/migrations/0019_remove_smeprojectmilestones_sme.py  1
- sme_management/migrations/0020_smeprojectmilestones_sme.py  1
- sme_management/tests/test_sme_apis.py  2
- sme_management/permissions.py  2
- sme_management/migrations/0012_auto_20200529_1437.py  1
- user_management/migrations/0003_auto_20200528_1650.py  1
- sme_management/views.py  4
         

Clones added
============
- sme_management/tests/test_sme_apis.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/BuildForSDG/Team-004-Backend/pullRequest?prid=5635476&bid=18455683)

File: sme_management/permissions.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        print(view)
+        return request.user.role == 'INVESTOR_USER' or request.user.role == 'BACKOFFICE_USER'
+
+
+class UploadProofForMilestone(permissions.BasePermission):
+    """Allow SME_USERs only upload proof for milestones they have access to"""
+
+    message = 'You do not have access to this milestone'
+
+    def has_object_permission(self, request, view, obj):
```
Comment (oyekanmiayo) [inline sme_management/permissions.py:26]: Codacy Issue found: [Method could be a function](https://app.codacy.com/gh/BuildForSDG/Team-004-Backend/pullRequest?prid=5635476)

File: investor_management/tests/test_investor_apis.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
+from django.test import TestCase
+from django.urls import reverse
+
+from rest_framework import status
+from rest_framework.test import APIClient
+
+from sme_management.tests.test_sme_apis import \
```
Comment (oyekanmiayo) [inline investor_management/tests/test_investor_apis.py:7]: Codacy Issue found: [Unused create_sme_user imported from sme_management.tests.test_sme_apis](https://app.codacy.com/gh/BuildForSDG/Team-004-Backend/pullRequest?prid=5635476)

File: sme_management/tests/test_sme_apis.py [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+
+    def test_unauthorized_patch_request_on_milestone_fails(self):
+        """
+        Test that an SME_USER cannot update a milestone that does
+        not belong to the SME_USER's SME
+        """
+        sme1 = create_sample_sme(org_name="Org 1", address="Addy 1")
+
+        # Create sme project with milestones for new user
+        sme_project, milestone = create_sample_sme_project_with_milestones(sme1)
```
Comment (oyekanmiayo) [inline sme_management/tests/test_sme_apis.py:298]: Codacy Issue found: [Unused variable 'sme_project'](https://app.codacy.com/gh/BuildForSDG/Team-004-Backend/pullRequest?prid=5635476)

File: sme_management/views.py [hunk 5]
```diff
@@ -2,7 +2,16 @@
 from rest_framework.permissions import IsAuthenticated
 from rest_framework import generics
 
+from rest_framework.parsers import FileUploadParser
+from rest_framework.response import Response
+from rest_framework.views import APIView
+from rest_framework import status, viewsets
+
+from django.db.models import Q
+from django.core.exceptions import ObjectDoesNotExist
```
Comment (oyekanmiayo) [inline sme_management/views.py]: Codacy Issue found: [Unused ObjectDoesNotExist imported from django.core.exceptions](https://app.codacy.com/gh/BuildForSDG/Team-004-Backend/pullRequest?prid=5635476)

File: sme_management/permissions.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
+from rest_framework import permissions
+from sme_management.models import SMEUser
+
+
+class ViewProjectTimeline(permissions.BasePermission):
+    """Allow only INVESTOR_USERs to views all projects that require financing"""
+
+    message = 'You do not have permission to see timeline'
+
+    def has_permission(self, request, view):
```
Comment (oyekanmiayo) [inline sme_management/permissions.py:10]: Codacy Issue found: [Method could be a function](https://app.codacy.com/gh/BuildForSDG/Team-004-Backend/pullRequest?prid=5635476)

File: sme_management/views.py [hunk 7]
```diff
@@ -23,3 +32,121 @@ class ManageSMEUserView(generics.RetrieveUpdateAPIView):
...
+            return Response(serializer.data, status=status.HTTP_201_CREATED)
+        else:
+            print(serializer.errors)
+            return Response(serializer.data, status=status.HTTP_400_BAD_REQUEST)
+
+    # TODO: In the future, we would want to make a separation between
+    #  how we get COMPLETED, IN_PROGRESS and UNAPPROVED projects
+    #  For now, we just return everything in one call
+    #  There might be a solution to this with search_fields. Confirm
+    def get(self, request, format=None):
```
Comment (oyekanmiayo) [inline sme_management/views.py]: Codacy Issue found: [Redefining built-in 'format'](https://app.codacy.com/gh/BuildForSDG/Team-004-Backend/pullRequest?prid=5635476)

File: sme_management/tests/test_sme_apis.py [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+        res = self.client.get(CREATE_SME_PROJECT_URL)
+
+        self.assertEqual(res.status_code, status.HTTP_200_OK)
+        for project in res.data:
+            self.assertNotEqual(project['id'], sme_project2.id)
+
+    def test_creating_milestones_for_project_is_successful(self):
+        """Test creating milestones for project is successful"""
+
+        payload, sme_project = get_create_milestone_payload()
```
Comment (oyekanmiayo) [inline sme_management/tests/test_sme_apis.py:249]: Codacy Issue found: [Unused variable 'sme_project'](https://app.codacy.com/gh/BuildForSDG/Team-004-Backend/pullRequest?prid=5635476)

**Review by nafiuadegbite — APPROVED** [review]

**Review by gblessylva — APPROVED** [review]

**Review by AceTheCreator — APPROVED** [review]

**Review by samoye01 — APPROVED** [review]
Good job man!

**Comment by oyekanmiayo** [comment]
> Good job man!

Thank you!

**Pull request merged**